<a href="https://colab.research.google.com/github/HarshESC/Deep-Learning/blob/main/Deep_Learning_Fashion_Minst.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:

import warnings

warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD



In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()
input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype("float32")
X_test = X_test.astype("float32")
X_train /= 255
X_test /= 255

y_train = to_categorical(y_train, nb_classes)
y_test = to_categorical(y_test, nb_classes)


11501568/11490434 [==============================] - 0s 0us/step


In [3]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))


In [4]:
model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])


In [5]:
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 1.2869 - accuracy: 0.6612
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5024 - accuracy: 0.8690
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3882 - accuracy: 0.8925
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3419 - accuracy: 0.9025
Epoch 5/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3137 - accuracy: 0.9109
Epoch 6/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2936 - accuracy: 0.9157
Epoch 7/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2777 - accuracy: 0.9208
Epoch 8/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2640 - accuracy: 0.9246
Epoch 9/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2522 - accuracy: 0.9282
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2418 - accuracy: 0.9307

In [6]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.1698606014251709
Test accuracy: 0.9521999955177307


In [7]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.7078 - accuracy: 0.4629
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.6752 - accuracy: 0.7937
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4375 - accuracy: 0.8712
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3477 - accuracy: 0.8979
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3025 - accuracy: 0.9114
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2717 - accuracy: 0.9201
Epoch 7/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2470 - accuracy: 0.9279
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2269 - accuracy: 0.9336
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2101 - accuracy: 0.9383
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.1958 - accuracy: 0.9426

In [8]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.12715891003608704
Test accuracy: 0.9607999920845032


In [9]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 6s 11ms/step - loss: 1.9243 - accuracy: 0.3727
Epoch 2/20
469/469 [==============================] - 5s 11ms/step - loss: 0.6496 - accuracy: 0.8115
Epoch 3/20
469/469 [==============================] - 5s 11ms/step - loss: 0.3891 - accuracy: 0.8882
Epoch 4/20
469/469 [==============================] - 5s 11ms/step - loss: 0.2950 - accuracy: 0.9164
Epoch 5/20
469/469 [==============================] - 5s 11ms/step - loss: 0.2430 - accuracy: 0.9308
Epoch 6/20
469/469 [==============================] - 5s 11ms/step - loss: 0.2066 - accuracy: 0.9410
Epoch 7/20
469/469 [==============================] - 5s 11ms/step - loss: 0.1798 - accuracy: 0.9488
Epoch 8/20
469/469 [==============================] - 5s 11ms/step - loss: 0.1603 - accuracy: 0.9546
Epoch 9/20
469/469 [==============================] - 5s 11ms/step - loss: 0.1433 - accuracy: 0.9597
Epoch 10/20
469/469 [==============================] - 5s 11ms/step - loss: 0.1303 - accura

In [10]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.10159134119749069
Test accuracy: 0.9692999720573425


In [11]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(1024, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 42s 89ms/step - loss: 2.0738 - accuracy: 0.3304
Epoch 2/20
469/469 [==============================] - 44s 93ms/step - loss: 0.7792 - accuracy: 0.7678
Epoch 3/20
469/469 [==============================] - 43s 93ms/step - loss: 0.3829 - accuracy: 0.8889
Epoch 4/20
469/469 [==============================] - 43s 92ms/step - loss: 0.2653 - accuracy: 0.9269
Epoch 5/20
469/469 [==============================] - 43s 91ms/step - loss: 0.2098 - accuracy: 0.9417
Epoch 6/20
469/469 [==============================] - 43s 92ms/step - loss: 0.1665 - accuracy: 0.9533
Epoch 7/20
469/469 [==============================] - 43s 91ms/step - loss: 0.1388 - accuracy: 0.9606
Epoch 8/20
469/469 [==============================] - 42s 90ms/step - loss: 0.1172 - accuracy: 0.9672
Epoch 9/20
469/469 [==============================] - 43s 92ms/step - loss: 0.1019 - accuracy: 0.9709
Epoch 10/20
469/469 [==============================] - 42s 89ms/step - loss: 0.085

In [12]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.11868144571781158
Test accuracy: 0.9700999855995178


The most accurate model came from using 9 layers. The speed of the model was much slower. The run time of a 7 layer model was noticeably faster but performed with a worse accuracy.



# Tanh activation and SGD loss


In [13]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(10, activation="softmax"))


In [14]:
model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])


In [15]:
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.0821 - accuracy: 0.7360
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5332 - accuracy: 0.8693
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4270 - accuracy: 0.8876
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3775 - accuracy: 0.8979
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3472 - accuracy: 0.9036
Epoch 6/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3257 - accuracy: 0.9090
Epoch 7/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3091 - accuracy: 0.9131
Epoch 8/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2954 - accuracy: 0.9164
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2839 - accuracy: 0.9200
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2736 - accuracy: 0.9226

In [16]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.20351992547512054
Test accuracy: 0.9422000050544739


In [17]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.3446 - accuracy: 0.6661
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.7702 - accuracy: 0.8376
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5641 - accuracy: 0.8742
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4588 - accuracy: 0.8913
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3959 - accuracy: 0.9019
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3536 - accuracy: 0.9097
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3223 - accuracy: 0.9160
Epoch 8/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2979 - accuracy: 0.9211
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 0.2779 - accuracy: 0.9256
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2611 - accuracy: 0.9300

In [18]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.17027774453163147
Test accuracy: 0.9498999714851379


In [19]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(256, activation="tanh"))
model.add(Dense(128, activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 5s 11ms/step - loss: 1.2148 - accuracy: 0.7202
Epoch 2/20
469/469 [==============================] - 5s 11ms/step - loss: 0.6218 - accuracy: 0.8772
Epoch 3/20
469/469 [==============================] - 5s 11ms/step - loss: 0.4403 - accuracy: 0.9023
Epoch 4/20
469/469 [==============================] - 5s 11ms/step - loss: 0.3568 - accuracy: 0.9151
Epoch 5/20
469/469 [==============================] - 5s 11ms/step - loss: 0.3073 - accuracy: 0.9235
Epoch 6/20
469/469 [==============================] - 5s 11ms/step - loss: 0.2728 - accuracy: 0.9306
Epoch 7/20
469/469 [==============================] - 5s 11ms/step - loss: 0.2470 - accuracy: 0.9371
Epoch 8/20
469/469 [==============================] - 5s 11ms/step - loss: 0.2256 - accuracy: 0.9414
Epoch 9/20
469/469 [==============================] - 5s 12ms/step - loss: 0.2076 - accuracy: 0.9460
Epoch 10/20
469/469 [==============================] - 5s 12ms/step - loss: 0.1918 - accura

In [20]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.12417525798082352
Test accuracy: 0.9650999903678894


In [21]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(1024, activation="tanh"))
model.add(Dense(512, activation="tanh"))
model.add(Dense(256, activation="tanh"))
model.add(Dense(128, activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 43s 91ms/step - loss: 0.9482 - accuracy: 0.8125
Epoch 2/20
469/469 [==============================] - 43s 92ms/step - loss: 0.4863 - accuracy: 0.9001
Epoch 3/20
469/469 [==============================] - 43s 91ms/step - loss: 0.3564 - accuracy: 0.9186
Epoch 4/20
469/469 [==============================] - 44s 94ms/step - loss: 0.2904 - accuracy: 0.9310
Epoch 5/20
469/469 [==============================] - 42s 90ms/step - loss: 0.2484 - accuracy: 0.9388
Epoch 6/20
469/469 [==============================] - 42s 89ms/step - loss: 0.2168 - accuracy: 0.9457
Epoch 7/20
469/469 [==============================] - 42s 90ms/step - loss: 0.1928 - accuracy: 0.9512
Epoch 8/20
469/469 [==============================] - 43s 92ms/step - loss: 0.1719 - accuracy: 0.9569
Epoch 9/20
469/469 [==============================] - 44s 94ms/step - loss: 0.1546 - accuracy: 0.9602
Epoch 10/20
469/469 [==============================] - 44s 93ms/step - loss: 0.139

In [22]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.09245689958333969
Test accuracy: 0.973800003528595


The model with 9 layers had the best test accuracy, however it was noticeably slower. Additionally, the accuracy was only better by around 0.01. Using tanh activation provided very similar results and the accuracy was just barely lower than when using ReLU.



# Sigmoid activation and SGD loss¶


In [23]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2785 - accuracy: 0.1497
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 2.2191 - accuracy: 0.3472
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 2.1490 - accuracy: 0.5147
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 2.0495 - accuracy: 0.5922
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 1.9096 - accuracy: 0.6316
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 1.7323 - accuracy: 0.6595
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 1.5395 - accuracy: 0.6862
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 1.3595 - accuracy: 0.7127
Epoch 9/20
469/469 [==============================] - 2s 3ms/step - loss: 1.2077 - accuracy: 0.7352
Epoch 10/20
469/469 [==============================] - 2s 3ms/step - loss: 1.0847 - accuracy: 0.7540

In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


In [26]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 2.2970941066741943
Test accuracy: 0.11349999904632568


In [ ]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(256, activation="sigmoid"))
model.add(Dense(128, activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(1024, activation="sigmoid"))
model.add(Dense(512, activation="sigmoid"))
model.add(Dense(256, activation="sigmoid"))
model.add(Dense(128, activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


The test accuracy for the sigmoid tests are all pretty low. It appears using sigmoid activation is not a good idea for this model.



#ReLU activation and SGD loss Changing Batch Sizes¶


In [ ]:
# Using 64 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Using 32 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=32, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Using 16 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=16, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Using 8 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=8, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


The score increases as the batch size decreases, however the run time increases as well. It was more noticeable when the batch size got to 16 and lower. The best score was obtained from using a batch size of 8, but it took more than twice the time.

Choosing between batch sizes, I would recommend using a batch size of 64 because it provides an accuracy that is almost just as good as having a batch size of 8 but also executing in less than half the time it took for the 8 batch size model to run.



# ReLU Activation, SGD Loss, Batchsize of 64, and Changing the Amount of Neurons¶


In [ ]:
# Having constant amount of 32 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(32, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Having constant amount of 64 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Having constant amount of 128 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Having constant amount of 256 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(256, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Having constant amount of 512 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


In [ ]:
# Having neurons decrease throughout model
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


In [ ]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


When using constant amounts of neurons the test accuracy increases as the neurons increase, however, once the neurons get too high the run time starts to take longer to execute. The models all appear to have similar accuracies, but the model with a constant amound of 256 neurons performs the best in the most optimal run time. It even has a better accuracy than the model with neurons that decrease within each layer.

# Conclusion
Overall, each model is able to reach a very high accuracy, usually over 0.95. models that achieve an accuracy higher than 0.97 start to take a lot longer to execute, and only have a small increase in accuracy. Additionally, whenever the layers were increased the accuracy would increase, however once the layers got past 7 the accuracy would barely increase and the run time would take much longer. Therefore, I figured the most optimal amount of layers would be 7.

It appears the best activation function was the ReLU activation. I was able to obtain an accuracy score of around 0.977 with a quick run time using a constant amound of 256 neurons, ReLU activation, 7 layers, and SGD loss function.

# Possible Next Steps¶
Some next steps that could be done for improving the models would be to change the loss functions (incorporate a hinge loss function instead of categorical). Additionally, I could change the optimizer to something other than sgd (adam).